adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1824,8 +1846,19 @@ protected final void close(final String reason, final boolean sendDisconnectedEv
          * To be called by the MqttEndpoint closeHandler.
          */
         protected final void onClose() {
-            final Span span = newSpan(stopCalled() ? "close device connection (server shutdown)" : "CLOSE");
-            final String reason = stopCalled() ? "server shutdown" : null;
+            final Span span;
+            final String reason;
+            if (protocolLevelException != null) {
+                span = newSpan("close connection due to protocol error");
+                TracingHelper.logError(span, protocolLevelException);
+                reason = "protocol error: " + protocolLevelException.toString();
+            } else if (stopCalled()) {
+                span = newSpan("close device connection (server shutdown)");
+                reason = "server shutdown";
+            } else {
+                span = newSpan("CLOSE");
+                reason = null;
+            }
```

Comment:
FMPOV using if - else if - else makes the code so much more readable than when using the conditional assignment :+1: